graph_storage.py

```diff
@@ -0,0 +1,98 @@
+import json
+from node import Node
+
+
+class GraphStorage:
+    json_file_name = 'all_nodes.json'
+
+    def __init__(self):
+        self.node_id_to_node = {}
+        with open(self.json_file_name, 'r') as all_nodes_json:
+            try:
+                json_object = json.loads(all_nodes_json)
+
+                if 'nodes' not in json_object:
+                    return
+
+                for node_dict in json_object['nodes']:
+                    self.node_id_to_node[node_dict['id']] = Node(node_dict)
+
+            except (TypeError, ValueError):
+                print('Exception, while reading json file.')
+                return
+
+    def _add_node(self, new_node_dict: dict):
+        node_to_process = self.node_id_to_node.get(new_node_dict['id'])
+        node_to_process_valid_parents = []
+        new_node_parent = new_node_dict.get('parent')
+        if node_to_process is None:
+            node_to_process = Node({
+                'id': new_node_dict['id'],
+                'parents': [] if new_node_parent is None else [new_node_parent]
+            })
+        elif node_to_process.parents and new_node_parent not in node_to_process.parents:
+            node_to_process_valid_parents = node_to_process.parents
+            node_to_process.parents += [new_node_parent]
+        else:
+            node_to_process.parents = [new_node_parent]
+            node_to_process.parents = list(filter(None.__ne__, node_to_process.parents))
+        if node_to_process.is_new_node_cyclic(self.node_id_to_node, new_node_parent):
+            node_to_process.parents = node_to_process_valid_parents
+            return False
+
+        if new_node_parent is not None and new_node_parent not in self.node_id_to_node.keys():
+            self.node_id_to_node[new_node_parent] = Node({'id': new_node_parent, 'parents': []})
+
+        self.node_id_to_node[node_to_process.node_id] = node_to_process
+        return True
+
+    def add_nodes(self, new_nodes_list: list):
+        cyclic_nodes = []
+        for new_node in new_nodes_list:
+            if not self._add_node(new_node):
+                cyclic_nodes.append(new_node)
+
+        with open('all_nodes.json', 'w') as all_nodes_json:
+            json.dump({'nodes': [node.__dict__ for node in self.node_id_to_node.values()]}, all_nodes_json)
+
+        return cyclic_nodes
+
+    def find_chains(self, node_id):
+        are_roots_reached = False
+        all_chains = []
+        node_id_related_chains = []
+        for youngest_node_id in list(self._find_youngest_nodes().keys()):
+            all_chains.append([youngest_node_id])
```

Comment:
Technically we are reading and then building all the chains that we have in our database and basically, it's extra work. We need build only that trees where our node id exists